controllers/openstackcontrolplane_controller.go

```diff
@@ -478,13 +463,135 @@ func (r *OpenStackControlPlaneReconciler) createOrGetTripleoPasswords(
 
 			return err
 		}
+	} else {
+		tripleoPasswordsByte := secret.Data["tripleo-overcloud-passwords.yaml"]
+
+		tripleoPasswordsRaw := make(map[string]interface{})
+		if err := yaml.Unmarshal(tripleoPasswordsByte, &tripleoPasswordsRaw); err != nil {
+			cond.Message = fmt.Sprintf("Error extract TripleoPasswords from Secret %s", controlplane.TripleoPasswordSecret)
+			cond.Reason = shared.ControlPlaneReasonTripleoPasswordsSecretError
+			cond.Type = shared.CommonCondTypeError
+			err = common.WrapErrorForObject(cond.Message, instance, err)
+
+			return err
+		}
+
+		// recursice convert nested map[interface{}]interface{} password data
+		// into map[string]interface{} and merge into a new tripleoPasswords map
+		tripleoPasswords, err := common.RecursiveMergeMaps(
+			map[string]interface{}{},
+			tripleoPasswordsRaw,
+			common.PasswordMaxDepth,
+		)
+		if err != nil {
+			cond.Message = fmt.Sprintf("Error recursive merge TripleoPasswords from Secret %s", controlplane.TripleoPasswordSecret)
+			cond.Reason = shared.ControlPlaneReasonTripleoPasswordsSecretError
+			cond.Type = shared.CommonCondTypeError
+			err = common.WrapErrorForObject(cond.Message, instance, err)
+
+			return err
+		}
+
+		// get actual password level of the password data, so everything bellow "parameter_defaults"
+		currentPasswordsMap := tripleoPasswords["parameter_defaults"].(map[string]interface{})
+
+		// If there are netsted maps bellow the highest level, Marshal them into a json string
+		for key, val := range currentPasswordsMap {
+			if val == nil {
+				cond.Message = fmt.Sprintf("Error, password for %s is nil!", key)
+				cond.Reason = shared.ControlPlaneReasonTripleoPasswordsSecretError
+				cond.Type = shared.CommonCondTypeError
+				err = common.WrapErrorForObject(cond.Message, instance, err)
+
+				return err
+			}
+
+			if common.IsInterfaceMap(val) {
+				b, err := json.Marshal(val)
+				if err != nil {
+					cond.Message = fmt.Sprintf("Error marshal nested password maps from TripleoPasswords from Secret %s", controlplane.TripleoPasswordSecret)
+					cond.Reason = shared.ControlPlaneReasonTripleoPasswordsSecretError
+					cond.Type = shared.CommonCondTypeError
+					err = common.WrapErrorForObject(cond.Message, instance, err)
+
+					return err
+				}
+
+				currentPasswordsMap[key] = string(b)
+			}
+		}
+
+		// validate if new password entries need to be added
+		newPasswordsMap := make(map[string]interface{})
+		for key := range common.GeneratePasswordsMap() {
```

Comment:
why not iterate passwordNames?

Reply:
right, will make it public so that it can be used here.